test/common/stats/thread_local_store_test.cc

```diff
@@ -1567,6 +1569,111 @@ TEST_F(HistogramTest, ParentHistogramBucketSummary) {
             parent_histogram->bucketSummary());
 }
 
+TEST_F(HistogramTest, ForEachHistogram) {
+  std::vector<std::reference_wrapper<Stats::Histogram>> histograms;
+
+  const size_t num_stats = 11;
+  for (size_t idx = 0; idx < num_stats; ++idx) {
+    auto stat_name = absl::StrCat("histogram.", idx);
+    histograms.emplace_back(
+        store_->histogramFromString(stat_name, Stats::Histogram::Unit::Unspecified));
+  }
+  EXPECT_EQ(histograms.size(), 11);
+
+  size_t num_histograms = 0;
+  size_t num_iterations = 0;
+  store_->forEachHistogram([&num_histograms](std::size_t size) { num_histograms = size; },
+                           [&num_iterations](Stats::ParentHistogram&) { ++num_iterations; });
+  EXPECT_EQ(num_histograms, 11);
+  EXPECT_EQ(num_iterations, 11);
+
+  Histogram& deleted_histogram = histograms[4];
+
+  // Verify that rejecting histograms removes them from the iteration set.
+  envoy::config::metrics::v3::StatsConfig stats_config_;
+  stats_config_.mutable_stats_matcher()->set_reject_all(true);
+  store_->setStatsMatcher(std::make_unique<StatsMatcherImpl>(stats_config_, symbol_table_));
+  num_histograms = 0;
+  num_iterations = 0;
+  store_->forEachHistogram([&num_histograms](std::size_t size) { num_histograms = size; },
+                           [&num_iterations](Stats::ParentHistogram&) { ++num_iterations; });
+  EXPECT_EQ(num_histograms, 0);
+  EXPECT_EQ(num_iterations, 0);
+
+  // Verify that we can access the local reference without a crash.
+  EXPECT_EQ(deleted_histogram.unit(), Stats::Histogram::Unit::Unspecified);
+}
+
+TEST_F(HistogramTest, ForEachSinkedHistogram) {
+  StatNamePool pool(store_->symbolTable());
+
+  std::unique_ptr<Stats::TestUtil::TestSinkPredicates> moved_sink_predicates =
+      std::make_unique<Stats::TestUtil::TestSinkPredicates>();
+  Stats::TestUtil::TestSinkPredicates* sink_predicates = moved_sink_predicates.get();
```

Comment:
this is not a good pattern -- I realize this pretty much has to work, but officially I'm pretty sure it's undefined behavior to save a pointer held in a unique_ptr that you then use as a the source of a move.

Instead, can you provide an accessor: Store::sinkPredicates() and reference them from the store, and change the other test-cases that use this pattern also?

Reply:
Added sinkPredicates() to StoreRoot.